diff --git a/init/upstart/boot-splash.conf b/init/upstart/boot-splash.conf
index 95e9d73a89..33c08ba748 100644
--- a/init/upstart/boot-splash.conf
+++ b/init/upstart/boot-splash.conf
@@ -25,7 +25,7 @@ script
     backlight_tool --set_brightness_percent=40.0 || true
   fi
 
-  BACKGROUND=0xfffefefe
+  BACKGROUND=0xff000000
   ASSETS=/usr/share/chromeos-assets
   ASSETS_200=${ASSETS}/images_200_percent
   ASSETS_100=${ASSETS}/images_100_percent
